lib/pages/search/public_room/search_public_room_controller.dart

```diff
@@ -0,0 +1,186 @@
+      ValueNotifier<PresentationSearchPublicRoomUIState>(
+    PresentationSearchPublicRoomInitial(),
+  );
+
+  static const int _limitPublicRoomSearchFilter = 20;
+
+  PublicRoomQueryFilter? _filter;
+
+  bool get searchTermIsNotEmpty =>
+      _filter?.genericSearchTerm?.isNotEmpty == true;
+
+  String? get genericSearchTerm => _filter?.genericSearchTerm;
+
+  void init() {
+    initializeDebouncer((keyword) {
+      _updateFilter(keyword);
+
+      if (keyword.isEmpty) return;
+
+      if (keyword.isRoomAlias()) {
+        _resetSearchResults();
+        _searchPublicRoom();
+      } else if (keyword.isRoomId()) {
+        _handleKeywordIsRoomId();
+      }
+    });
+  }
+
+  void _updateFilter(String keyword) {
+    _filter = PublicRoomQueryFilter(
+      genericSearchTerm: keyword,
+    );
+  }
+
+  void _resetSearchResults() {
+    searchResultsNotifier.value = PresentationSearchPublicRoom(
+      searchResults: [],
+    );
+  }
+
+  void _searchPublicRoom() {
+    _publicRoomInteractor
+        .execute(
+          filter: _filter,
+          limit: _limitPublicRoomSearchFilter,
+          server: getServerName(_filter?.genericSearchTerm),
+        )
+        .listen(
+          (searchResult) => _handleListenSearchPublicRoom(searchResult),
+        );
+  }
+
+  void _handleListenSearchPublicRoom(Either<Failure, Success> searchResult) {
+    searchResult.fold((failure) => _resetSearchResults(), (success) {
+      if (!searchTermIsNotEmpty) {
+        return;
+      }
+
+      if (success is PublicRoomSuccess) {
+        if (success.publicRoomsChunk == null) {
+          searchResultsNotifier.value = PresentationSearchPublicRoomEmpty();
+        } else if (success.publicRoomsChunk!.isEmpty) {
+          searchResultsNotifier.value = PresentationSearchPublicRoomEmpty();
+        } else {
+          searchResultsNotifier.value = PresentationSearchPublicRoom(
+            searchResults: success.publicRoomsChunk!,
+          );
+        }
+      }
+    });
+  }
+
+  void _handleKeywordIsRoomId() {
+    searchResultsNotifier.value = PresentationSearchPublicRoomEmpty();
+  }
+
+  void _viewRoom(BuildContext context, String roomId) {
+    context.go('/rooms/$roomId');
+  }
+
+  String? getServerName(String? roomIdOrAlias) {
+    if (roomIdOrAlias != null) {
+      return roomIdOrAlias.getServerNameFromRoomIdOrAlias();
+    }
+    return null;
+  }
+
+  void joinRoom(
+    BuildContext context,
+    String roomIdOrAlias,
+    String? server,
+  ) async {
+    final client = Matrix.of(context).client;
+    final result = await TwakeDialog.showFutureLoadingDialogFullScreen(
+      future: () => client.joinRoom(
+        roomIdOrAlias,
+        serverName: server != null ? [server] : null,
+      ),
+    );
+    if (result.error == null) {
+      if (client.getRoomById(result.result!) == null) {
+        await client.onSync.stream.firstWhere(
+          (sync) => sync.rooms?.join?.containsKey(result.result) ?? false,
+        );
```

Comment:
dispose when not used

Reply:
IMO, when using firstWhere method on a Stream, it automatically cancels the subscription.